content/blog/2025-06-15-optimizing-sql-dataframes-part-one.md

```diff
@@ -0,0 +1,250 @@
+`WHAT` you want to compute from the `HOW` to do it. SQL is a *declarative*
+language—it describes what answers are desired rather than an *imperative*
+language such as Python, where you describe how to do the computation as shown
+in Figure 1.
+
+<img src="/blog/images/optimizing-sql-dataframes/query-execution.png" width="80%" class="img-responsive" alt="Fig 1: Query Execution."/>
+
+**Figure 1**: Query Execution: Users describe the answer they want using either
+SQL or a DataFrame. For SQL, a Query Planner translates the parsed query 
+into an *initial plan*. The DataFrame API creates an initial plan directly.
+The initial plan is correct, but slow. Then, the Query
+Optimizer rewrites the initial plan into an *optimized plan*, which computes
+the same results but faster and more efficiently. Finally, the Execution Engine
+executes the optimized plan producing results.
+
+## SQL, DataFrames, LogicalPlan Equivalence
+
+Given their name, it is not surprising that Query Optimizers can improve the
+performance of SQL queries. However, it is under-appreciated that this also
+applies to DataFrame style APIs.
+
+Classic DataFrame systems such as [pandas] and [Polars] (by default) execute
+eagerly and thus have limited opportunities for optimization. However, more
+modern APIs such as [Polar's lazy API], [Apache Spark's DataFrame]. and
+[DataFusion's DataFrame] are much faster as they use the design shown in Figure
+1 and apply many query optimization techniques.
+
+[pandas]: https://pandas.pydata.org/
+[Polars]: https://pola.rs/) 
+[Polar'’'s lazy API]: https://docs.pola.rs/user-guide/lazy/using/
+[Apache Spark's DataFrame]: https://spark.apache.org/docs/latest/sql-programming-guide.html#datasets-and-dataframes),
+[DataFusion's DataFrame]: https://datafusion.apache.org/user-guide/dataframe.html
+
+## Example of Query Optimizer
+
+This section motivates the value of a Query Optimizer with an example. Let’s say
+you have some observations of animal behavior, as illustrated in Table 1.
+
+<img src="/blog/images/optimizing-sql-dataframes/table1.png" width="75%" class="img-responsive" alt="Table 1: Observational Data."/>
+
+**Table 1**: Example observational data.
+
+If the user wants to know the average population for some species in the last
+month, a user can write a SQL query or a DataFrame such as the following:
+
+SQL:
+
+```sql
+SELECT location, AVG(population)
+FROM observations
+WHERE species = ‘contrarian spider’ AND 
+  observation_time >= now() - interval '1 month'
+GROUP BY location
+```
+
+DataFrame:
+
+```rust
+df.scan("observations")
+  .filter(col("species").eq("contrarian spider"))
+  .filter(col("observation_time").ge(now()).sub(interval('1 month')))
+  .agg(vec![col(location)], vec![avg(col("population")])
+```
+
+Within DataFusion, both the SQL and DataFrame are translated into the same
+[`LogicalPlan`] , a “tree of relational operators.” This is a fancy way of
```

Comment:
```suggestion
[LogicalPlan], a “tree of relational operators.” This is a fancy way of
```

i dont think we can do both code highlight (backtick) and hyperlink

Reply:
![Screenshot 2025-06-12 at 8 53 33 AM](https://github.com/user-attachments/assets/886501da-decf-421f-8276-3d3f8fde27d4)